programs/08-wasm-eth/src/contract.rs

```diff
@@ -0,0 +1,649 @@
+        SudoMsg::VerifyMembership(verify_membership_msg) => {
+            verify_membership(deps.as_ref(), verify_membership_msg)?
+        }
+        SudoMsg::VerifyNonMembership(verify_non_membership_msg) => {
+            verify_non_membership(deps.as_ref(), verify_non_membership_msg)?
+        }
+        SudoMsg::UpdateState(_) => update_state()?,
+        SudoMsg::UpdateStateOnMisbehaviour(_) => unimplemented!(),
+        SudoMsg::VerifyUpgradeAndUpdateState(_) => unimplemented!(),
+        SudoMsg::MigrateClientStore(_) => unimplemented!(),
+    };
+
+    Ok(Response::default().set_data(result))
+}
+
+pub fn verify_membership(
+    deps: Deps<EthereumCustomQuery>,
+    verify_membership_msg: VerifyMembershipMsg,
+) -> Result<Binary, ContractError> {
+    let eth_client_state = get_eth_client_state(deps);
+    let eth_consensus_state = get_eth_consensus_state(deps, &verify_membership_msg.height);
+
+    ethereum_light_client::membership::verify_membership(
+        eth_consensus_state,
+        eth_client_state,
+        verify_membership_msg.proof.into(),
+        verify_membership_msg
+            .merkle_path
+            .key_path
+            .into_iter()
+            .map(Into::into)
+            .collect(),
+        Some(verify_membership_msg.value.into()),
+    )
+    .map_err(ContractError::VerifyMembershipFailed)?;
+
+    Ok(to_json_binary(&Ok::<(), ()>(()))?)
+}
+
+pub fn verify_non_membership(
+    deps: Deps<EthereumCustomQuery>,
+    verify_non_membership_msg: VerifyNonMembershipMsg,
+) -> Result<Binary, ContractError> {
+    let eth_client_state = get_eth_client_state(deps);
+    let eth_consensus_state = get_eth_consensus_state(deps, &verify_non_membership_msg.height);
+
+    ethereum_light_client::membership::verify_membership(
+        eth_consensus_state,
+        eth_client_state,
+        verify_non_membership_msg.proof.into(),
+        verify_non_membership_msg
+            .merkle_path
+            .key_path
+            .into_iter()
+            .map(Into::into)
+            .collect(),
+        None,
+    )
+    .map_err(ContractError::VerifyNonMembershipFailed)?;
+
+    Ok(to_json_binary(&Ok::<(), ()>(()))?)
+}
+
+pub fn update_state() -> Result<Binary, ContractError> {
+    Ok(to_json_binary(&UpdateStateResult { heights: vec![] })?)
+}
+
+#[entry_point]
+pub fn execute(
+    _deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    _msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    unimplemented!()
+}
+
+#[entry_point]
+pub fn query(
+    deps: Deps<EthereumCustomQuery>,
+    env: Env,
+    msg: QueryMsg,
+) -> Result<Binary, ContractError> {
+    match msg {
+        QueryMsg::VerifyClientMessage(verify_client_message_msg) => {
+            verify_client_message(deps, env, verify_client_message_msg)
+        }
+        QueryMsg::CheckForMisbehaviour(_) => check_for_misbehaviour(),
+        QueryMsg::TimestampAtHeight(_) => timestamp_at_height(env),
+        QueryMsg::Status(_) => status(),
```

Comment:
These should all be moved to a `query` submodule
```suggestion
        QueryMsg::VerifyClientMessage(verify_client_message_msg) => {
            query::verify_client_message(deps, env, verify_client_message_msg)
        }
        QueryMsg::CheckForMisbehaviour(_) => query::check_for_misbehaviour(),
        QueryMsg::TimestampAtHeight(_) => query::timestamp_at_height(env),
        QueryMsg::Status(_) => query::status(),
```

Reply:
Yes, but not necessary in this PR. The wasm client itself will be cleaned up and made proper in https://github.com/cosmos/solidity-ibc-eureka/issues/143